File: Lab1/cpp/TMixFraction.cpp
#include "TMixFraction.h"

TMixFraction::TMixFraction() {
    this->mix = 0;
    this->upNumber = 0;
    this->downNumber = 1;
}

TMixFraction::TMixFraction(int mix, int up, int down) {
    this->mix=mix;
    this->upNumber=up;
    this->downNumber=down;
}

TMixFraction::TMixFraction(const TMixFraction &copy)  : TFraction(copy) {
    this->mix=copy.mix;
    this->upNumber=copy.upNumber;
    this->downNumber=copy.downNumber;
}

int TMixFraction::getMixNumber() const {
    return this->mix;
}

void TMixFraction::inputFraction() {
    try {
        cout<<" Enter numerator: ";
        cin >> this->upNumber;
        cout<<"Enter denominator: ";
        cin>> this->downNumber;
    }
    catch (...)
    {
        cout<<"Wrong format of data: ";
    }
}

void TMixFraction::printFraction() {
    cout << "Mix: " + to_string(mix) + " Fraction: " + to_string(upNumber) + "/" + to_string(downNumber);
}

void TMixFraction::simplifyFraction() {
    this->upNumber = (this->mix * this->downNumber) + this->upNumber;
    int denominator = findGCD(this->upNumber, this->downNumber);
    this->upNumber /= denominator;
    this->downNumber /= denominator;
    int upLeft = this->upNumber % this->downNumber;
    this->mix = (this->upNumber - upLeft) / this->downNumber;
    this->upNumber = upLeft;
}

TMixFraction TMixFraction::operator+(const TMixFraction &obj) {
    this->upNumber = (this->mix * this->downNumber) + this->upNumber;
    int objUpNumber = (obj.mix * obj.downNumber) + obj.upNumber;
    int lcm = (this->downNumber*obj.downNumber)/findGCD(this->downNumber,obj.downNumber);
    int sum=(this->upNumber*lcm/this->downNumber) + (objUpNumber*lcm/obj.downNumber);
    int sumLeft = sum % lcm;
    int mixNumber = (sum - sumLeft) / lcm;
    return {mixNumber, sumLeft, lcm};
}

TMixFraction TMixFraction::operator-(const TMixFraction &obj) {
    this->upNumber = (this->mix * this->downNumber) + this->upNumber;
    int objUpNumber = (obj.mix * obj.downNumber) + obj.upNumber;
    int lcm = (this->downNumber*obj.downNumber)/findGCD(this->downNumber,obj.downNumber);
    int sum=(this->upNumber*lcm/this->downNumber) - (objUpNumber*lcm/obj.downNumber);
    int sumLeft = sum % lcm;
    int mixNumber = (sum - sumLeft) / lcm;
    return {mixNumber, sumLeft, lcm};
}

TMixFraction TMixFraction::operator*(const TMixFraction &obj) {
    this->upNumber = (this->mix * this->downNumber) + this->upNumber;
    int objUpNumber = (obj.mix * obj.downNumber) + obj.upNumber;
    int sum = this->upNumber*objUpNumber;
    int lcm = this->downNumber*obj.downNumber;
    int sumLeft = sum%lcm;
    int mixNumber = (sum-sumLeft)/lcm;
    return { mixNumber, sumLeft, lcm};
}

TMixFraction TMixFraction::operator/(const TMixFraction &obj) {
    if(obj.downNumber==0){
        cout<<"Can't divide by 0";
        return {};
    }
    this->upNumber = (this->mix * this->downNumber) + this->upNumber;
    int objUpNumber = (obj.mix * obj.downNumber) + obj.upNumber;
    int sum = this->upNumber*obj.downNumber;
    int lcm = this->downNumber*objUpNumber;
    int sumLeft = sum%lcm;
    int mixNumber = (sum-sumLeft)/lcm;
    return { mixNumber, sumLeft, lcm};
}

File: Lab1/csharp/TMixFraction.cs
namespace csharp {
    public class TMixFraction : TFraction
    {
        private int _mix;

        public int MixNumber
        {
            get => _mix;
            set => _mix = value;
        }

        public TMixFraction()
        {
            this._mix = 0;
            this._upNumber = 0;
            this._downNumber = 1;
        }

        public TMixFraction(int mix, int up, int down)
        {
            this._mix = mix;
            this._upNumber = up;
            this._downNumber = down;
        }

        public TMixFraction(in TMixFraction copy) : base(copy)
        {
            this._mix = copy._mix;
            this._upNumber = copy._upNumber;
            this._downNumber = copy._downNumber;
        }

        public override void InputFraction()
        {
            try
            {
                Console.Write("Enter mix: ");
                this._mix = int.Parse(Console.ReadLine() ?? "0");
                Console.Write("Enter numerator: ");
                this._upNumber = int.Parse(Console.ReadLine() ?? "0");
                Console.Write("Enter denominator: ");
                this._downNumber = int.Parse(Console.ReadLine() ?? "0");
            }
            catch
            {
                Console.Write("Wrong format of data: ");
                Console.Write("\n");
            }
        }

        public override void PrintFraction()
        {
            Console.Write("Mix: " + Convert.ToString(_mix) + " Fraction: " + Convert.ToString(_upNumber) + "/" +
                Convert.ToString(_downNumber));
            Console.Write("\n");
        }

        public override void SimplifyFraction()
        {
            this._upNumber = (this._mix * this._downNumber) + this._upNumber;
            int denominator = FindGCD(this._upNumber, this._downNumber);
            if (denominator == 0)
            {
                return;
            }

            this._upNumber /= denominator;
            this._downNumber /= denominator;
            int upLeft = this._upNumber % this._downNumber;
            this._mix = (this._upNumber - upLeft) / this._downNumber;
            this._upNumber = upLeft;
        }

        public static TMixFraction operator +(TMixFraction impliedObject, in TMixFraction obj)
        {
            impliedObject._upNumber = (impliedObject._mix * impliedObject._downNumber) + impliedObject._upNumber;
            int objUpNumber = (obj._mix * obj._downNumber) + obj._upNumber;
            int lcm = (impliedObject._downNumber * obj._downNumber) / FindGCD(impliedObject._downNumber, obj._downNumber);
            int sum = (impliedObject._upNumber * lcm / impliedObject._downNumber) + (objUpNumber * lcm / obj._downNumber);
            int sumLeft = sum % lcm;
            int mixNumber = (sum - sumLeft) / lcm;
            return new TMixFraction(mixNumber, sumLeft, lcm);
        }

        public static TMixFraction operator -(TMixFraction impliedObject, in TMixFraction obj)
        {
            impliedObject._upNumber = (impliedObject._mix * impliedObject._downNumber) + impliedObject._upNumber;
            int objUpNumber = (obj._mix * obj._downNumber) + obj._upNumber;
            int lcm = (impliedObject._downNumber * obj._downNumber) / FindGCD(impliedObject._downNumber, obj._downNumber);
            int sum = (impliedObject._upNumber * lcm / impliedObject._downNumber) - (objUpNumber * lcm / obj._downNumber);
            int sumLeft = sum % lcm;
            int mixNumber = (sum - sumLeft) / lcm;
            return new TMixFraction(mixNumber, sumLeft, lcm);
        }

        public static TMixFraction operator *(TMixFraction impliedObject, in TMixFraction obj)
        {
            impliedObject._upNumber = (impliedObject._mix * impliedObject._downNumber) + impliedObject._upNumber;
            int objUpNumber = (obj._mix * obj._downNumber) + obj._upNumber;
            int sum = impliedObject._upNumber * objUpNumber;
            int lcm = impliedObject._downNumber * obj._downNumber;
            int sumLeft = sum % lcm;
            int mixNumber = (sum - sumLeft) / lcm;
            return new TMixFraction(mixNumber, sumLeft, lcm);
        }

        public static TMixFraction operator /(TMixFraction impliedObject, in TMixFraction obj)
        {
            if (obj._downNumber == 0)
            {
                Console.Write("Can't divide by 0 \n");
                return new TMixFraction();
            }

            impliedObject._upNumber = (impliedObject._mix * impliedObject._downNumber) + impliedObject._upNumber;
            int objUpNumber = (obj._mix * obj._downNumber) + obj._upNumber;
            int sum = impliedObject._upNumber * obj._downNumber;
            int lcm = impliedObject._downNumber * objUpNumber;
            int sumLeft = sum % lcm;
            int mixNumber = (sum - sumLeft) / lcm;
            return new TMixFraction(mixNumber, sumLeft, lcm);
        }
    }
}
